src/cdk/scrolling/virtual-scroll-viewport.ts

```diff
@@ -0,0 +1,189 @@
+      this._ngZone.run(() => {
+        this._totalContentSize = size;
+        this._changeDetectorRef.markForCheck();
+      });
+    }
+  }
+  private _totalContentSize = 0;
+
+  /** The currently rendered range of indices. */
+  get renderedRange() { return this._renderedRange; }
+  set renderedRange(range: Range) {
+    if (!this._rangesEqual(this._renderedRange, range)) {
+      this._ngZone.run(() => {
+        this._renderedRangeSubject.next(this._renderedRange = range);
+        this._changeDetectorRef.markForCheck();
+      });
+    }
+  }
+  private _renderedRange: Range = {start: 0, end: 0};
+
+  /** The offset of the rendered portion of the data from the start. */
+  get renderedContentOffset() { return this._renderedContentOffset; }
+  set renderedContentOffset(offset) {
+    if (this._renderedContentOffset != offset) {
+      this._ngZone.run(() => {
+        this._renderedContentOffset = offset;
+        this._renderedContentTransform = this.orientation === 'horizontal' ?
+            `translateX(${offset}px)`: `translateY(${offset}px)`;
+        this._changeDetectorRef.markForCheck();
+      });
+    }
+  }
+  private _renderedContentOffset = 0;
+
+  /** The length of the data connected to this viewport. */
+  get dataLength() { return this._dataLength; }
+  private _dataLength = 0;
+
+  /** The size of the viewport. */
+  get viewportSize() { return this._viewportSize; }
+  private _viewportSize = 0;
+
+  /** A stream that emits whenever the rendered range changes. */
+  renderedRangeStream: Observable<Range> = this._renderedRangeSubject.asObservable();
+
+  _renderedContentTransform: string;
+
+  private _connected = false;
+
+  private _scrollHandledStatus: 'needed' | 'pending' | 'done' = 'done';
+
+  private _scrollStrategyInited = false;
+
+  constructor(public elementRef: ElementRef, private _changeDetectorRef: ChangeDetectorRef,
+              private _ngZone: NgZone,
+              @Inject(VIRTUAL_SCROLL_STRATEGY) private _scrollStrategy: VirtualScrollStrategy) {}
+
+  /** Connect a `CdkForOf` to this viewport. */
+  connect(forOf: CdkForOf<any>) {
+    if (this._connected) {
+      throw Error('CdkVirtualScrollViewport is already connected.');
+    }
+
+    this._connected = true;
+    forOf.dataStream.pipe(takeUntil(this._disconnectSubject)).subscribe(data => {
+      const len = data.length;
+      if (len != this._dataLength) {
+        this._dataLength = len;
+        if (this._scrollStrategyInited) {
+          this._scrollStrategy.onDataLengthChanged();
+        }
+      }
+    });
+  }
+
+  /** Disconnect the current `CdkForOf`. */
+  disconnect() {
+    this._connected = false;
+    this._disconnectSubject.next();
+  }
+
+  /** Gets the current scroll offset of the viewport. */
+  measureScrollOffset() {
+    return this.orientation === 'horizontal' ?
+        this.elementRef.nativeElement.scrollLeft : this.elementRef.nativeElement.scrollTop;
+  }
+
+  ngOnInit() {
+    Promise.resolve().then(() => {
```

Comment:
Rather than wrapping it in a promise, have you tried running it in one of the later lifecycle hooks?

Reply:
I tried, it seemed like they were not late enough and I had to use the promise